File: LeetCode/Stack/python/minStack.py
from typing import Dict, List


class MinStack:

    def __init__(self):
        self.stack = []
        self.mStack = []
        
    def push(self, val: int) -> None:
        self.stack.append(val)
        tmpVal = min(val, self.mStack[-1] if len(self.mStack) > 0 else val)
        self.mStack.append(tmpVal)

    def pop(self) -> None:
        self.stack.pop()
        self.mStack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.mStack[-1]
        
    @staticmethod
    def testSolution(input: Dict[str, List[str] | List[List[int]]]) -> None:
        print("Input: operation: {}".format(input["operation"]))
        print("Input: values: {}".format(input["values"]))
        print("Expected: {}".format(input["expected"]))
        output = []
        soln = None

        for i in range(0, len(input["operation"])):
            oper = input["operation"][i]
            if oper == "MinStack":
                output.append("null")
                soln = MinStack()
            elif oper == "push":
                output.append("null")
                soln.push(input["values"][i][0])
            elif oper == "pop":
                output.append("null")
                soln.pop()
            elif oper == "top":
                val = soln.top()
                output.append(str(val))
            elif oper == "getMin":
                val = soln.getMin()
                output.append(str(val))

        print("Result: {}".format(output))
        res = MinStack.validateResult(output, input["expected"])
        print("PASS" if res else "FAIL")
        print("-" * 50)

    @staticmethod
    def validateResult(output: List[str], expected: List[str]) -> bool:
        if len(output) != len(expected):
            return False
        
        for i in range(len(output)):
            if output[i] != expected[i]:
                return False
        
        return True

# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

if __name__ == "__main__":
    input = { "operation" : ["MinStack","push","push","push","getMin","pop","top","getMin"],
             "values": [[],[-2],[0],[-3],[],[],[],[]],
             "expected": ["null","null","null","null","-3","null","0","-2"]}
    
    MinStack.testSolution(input)

File: LeetCode/Stack/java/minStack.java
package LeetCode.Stack.java;

import java.util.Arrays;
import java.util.Stack;

record minStackRecord(String[] operation, int[][] val, String[] expected) {
}

public class minStack {
    private Stack<Integer> stack;
    private Stack<Integer> mStack;

    public minStack() {
        this.stack = new Stack<>();
        this.mStack = new Stack<>();
    }

    public void push(int val) {
        this.stack.push(val);
        int tmpVal = Math.min(val, !this.mStack.isEmpty() ? this.mStack.peek() : val);
        this.mStack.push(tmpVal);
    }

    public void pop() {
        this.stack.pop();
        this.mStack.pop();
    }

    public int top() {
        return this.stack.peek();
    }

    public int getMin() {
        return this.mStack.peek();
    }

    public static void main(String[] args) {
        minStackRecord input = new minStackRecord(
                new String[] { "MinStack", "push", "push", "push", "getMin", "pop", "top", "getMin" },
                new int[][] { {}, { -2 }, { 0 }, { -3 }, {}, {}, {}, {} },
                new String[] { "null", "null", "null", "null", "-3", "null", "0", "-2" });

        minStack.testSolution(input);
    }

    private static void testSolution(minStackRecord input) {
        System.out.println("Input: Operations: " + Arrays.toString(input.operation()) + "\n\t"
                + " values: " + Arrays.deepToString(input.val()));

        int n = input.operation().length;
        String[] output = new String[n];
        minStack soln = null;

        for (int i = 0; i < n; i += 1) {
            switch (input.operation()[i].strip()) {
                case "MinStack" -> {
                    soln = new minStack();
                    output[i] = "null";
                }
                case "push" -> {
                    output[i] = "null";
                    soln.push(input.val()[i][0]);
                }
                case "pop" -> {
                    output[i] = "null";
                    soln.pop();
                }
                case "top" -> {
                    output[i] = String.valueOf(soln.top());
                }
                case "getMin" -> {
                    output[i] = String.valueOf(soln.getMin());
                }
            }
        }

        System.out.println("Expected: " + Arrays.toString(input.expected()));
        System.out.println("Result: " + Arrays.toString(output));

        boolean res = minStack.validateResult(output, input.expected());
        System.out.println(res ? "PASS" : "FAIL");
    }

    private static boolean validateResult(String[] output, String[] expected) {
        if (output.length != expected.length) {
            return false;
        }

        for (int i = 0; i < output.length; i++) {
            if (!output[i].equals(expected[i])) {
                return false;
            }
        }

        return true;
    }
}
